File: Day 12/Solution.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        # Using two pointer approach with different speeds
        slow = nums[0]
        fast = nums[0]

        # Move slow by one step and fast by two steps...
        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break

        # Move slow to the start and move both pointers by one step ....
        slow = nums[0]
        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]

        return slow

File: Day 12/Solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findDuplicate = function(nums) {
    let slow = nums[0];
    let fast = nums[0];

    while (true) {
        slow = nums[slow];
        fast = nums[nums[fast]];
        if (slow === fast) {
            break;
        }
    }

    slow = nums[0];
    while (slow !== fast) {
        slow = nums[slow];
        fast = nums[fast];
    }

    return slow;
};